src/main/java/org/jabref/gui/cleanup/CleanupAction.java

```diff
@@ -135,17 +144,41 @@ private void showResults() {
     }
 
     private void cleanup(BibDatabaseContext databaseContext, CleanupPreferences cleanupPreferences) {
+        this.resetCleanUpFailures();
+        NamedCompound ce = new NamedCompound(Localization.lang("Cleanup entries"));
+
         for (BibEntry entry : stateManager.getSelectedEntries()) {
             // undo granularity is on entry level
-            NamedCompound ce = new NamedCompound(Localization.lang("Cleanup entry"));
-
             doCleanup(databaseContext, cleanupPreferences, entry, ce);
 
-            ce.end();
             if (ce.hasEdits()) {
                 modifiedEntriesCount++;
-                undoManager.addEdit(ce);
             }
         }
+
+        ce.end();
+
+        if (ce.hasEdits()) {
+            undoManager.addEdit(ce);
+        }
+
+        if (!failures.isEmpty()) {
+            // Show errors to user through dialog service
+            showFailures(failures);
+        }
+    }
+
+    private void showFailures(List<JabRefException> failures) {
+        StringBuilder sb = new StringBuilder();
+        for (JabRefException exception : failures) {
+            sb.append("- ").append(exception.getLocalizedMessage()).append("\n");
+        }
```

Comment:
Instead of `StringBuilder`, `StringJoinger` is the more modern Java approach. Even more modern is to use `.stream().collect(Collectors.joining(...)`. Please try out this.